### PR TITLE
[foundationdb] add a new client for FoundationDB

### PR DESCRIPTION
**Pull request by dbjoa**
Here is a new client for FoundationDB, which is based on the API version 5.1.7.

### CODE REVIEW

**Comment by dbjoa** [comment]
In order to test the binding, you need to download the official java client from the links below and install them manually to your local maven repo.

https://www.foundationdb.org/downloads/5.1.7/bindings/java/fdb-java-5.1.7.jar
https://www.foundationdb.org/downloads/5.1.7/bindings/java/fdb-java-5.1.7-javadoc.jar

``$ mvn install:install-file -Dfile=fdb-java-5.1.7.jar -DgroupId=com.apple.foundationdb -DartifactId=fdb-java -Dversion=5.1.7 -Dpackaging=jar``

**Comment by busbey** [comment]
is there an issue filed with the foundationDB folks already about publishing to maven central or otherwise adding a repository?

**Comment by dbjoa** [comment]
@busbey 
Yes, the issue has been filed already: https://github.com/apple/foundationdb/issues/219

**Comment by busbey** [comment]
thanks for the pointer! we'll block on that issue closing out then.

**Comment by busbey** [comment]
foundation db have published client jars to maven central now. FYI, @dbjoa. I'll see if restarting the travis build is enough.

**Comment by dbjoa** [comment]
@busbey I've updated the PR.

File: foundationdb/pom.xml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    <artifactId>binding-parent</artifactId>
+    <version>0.15.0-SNAPSHOT</version>
+    <relativePath>../binding-parent</relativePath>
+  </parent>
+
+  <artifactId>foundationdb-binding</artifactId>
+  <name>FoundationDB Binding</name>
+  <packaging>jar</packaging>
+  <repositories>
+    <repository>
```
Comment (busbey) [inline foundationdb/pom.xml]: I don't think we need the snapshot repository. can we drop it?

File: foundationdb/src/main/java/com/yahoo/ycsb/db/package-info.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+/**
+ * The YCSB binding for <a href="https://www.foundationdb.org">FoundationDB</a>.
+ */
+package com.yahoo.ycsb.db;
```
Comment (busbey) [inline foundationdb/src/main/java/com/yahoo/ycsb/db/package-info.java]: please put this into a dedicated package e.g. `com.yahoo.ycsb.db.foundationdb`

File: foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+      String apiVersion = props.getProperty(API_VERSION, API_VERSION_DEFAULT);
+      String clusterFile = props.getProperty(CLUSTER_FILE, CLUSTER_FILE_DEFAULT);
+      String dbBatchSize = props.getProperty(DB_BATCH_SIZE, DB_BATCH_SIZE_DEFAULT);
+      dbName = props.getProperty(DB_NAME, DB_NAME_DEFAULT);
+
+      logger.info("API Version: {}", apiVersion);
+      logger.info("Cluster File: {}\n", clusterFile);
+
+      try {
+        fdb = FDB.selectAPIVersion(Integer.parseInt(apiVersion.trim()));
```
Comment (busbey) [inline foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java]: is this the bit that needs to be protected by the synchronized block? is the db object itself threadsafe?

File: foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+  public void cleanup() throws DBException {
+    synchronized (INIT_COUNT) {
+      final int curInitCount = INIT_COUNT.decrementAndGet();
+      if (batchCount > 0) {
+        batchInsert();
+        batchCount = 0;
+      }
+      if (curInitCount == 0) {
+        try {
+          db.close();
```
Comment (busbey) [inline foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java]: over in `init()` we did fdb.open once for each client thread, but here we're only doing a single close. is that correct?

File: foundationdb/pom.xml (hunk 1 above)
Comment (dbjoa) [reply foundationdb/pom.xml]: Thank you for the comment. I've dropped the snapshot repository from the PR.

File: foundationdb/src/main/java/com/yahoo/ycsb/db/package-info.java (hunk 2 above)
Comment (dbjoa) [reply foundationdb/src/main/java/com/yahoo/ycsb/db/package-info.java]: Thank you for your suggestion. I've made the dedicated package.

File: foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java (hunk 3 above)
Comment (dbjoa) [reply foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java]: Thank you for the comment. The synchronized block is not needed. I've removed it from the PR.

File: foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java (hunk 4 above)
Comment (dbjoa) [reply foundationdb/src/main/java/com/yahoo/ycsb/db/FoundationDBClient.java]: Thank you for your detailed comment. db.close() should be called for each client thread. I've updated the PR.

**Comment by dbjoa** [comment]
@busbey I've changed the PR.

File: foundationdb/pom.xml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    <relativePath>../binding-parent</relativePath>
+  </parent>
+
+  <artifactId>foundationdb-binding</artifactId>
+  <name>FoundationDB Binding</name>
+
+  <packaging>jar</packaging>
+  <properties>
+    <!-- Tests do not run on jdk9 -->
+    <skipJDK9Tests>true</skipJDK9Tests>
```
Comment (busbey) [inline foundationdb/pom.xml]: Sorry for the churn, but a PR just landed that adds JDK10 and JDK11 test runs. Could you add the additional skip properties for those?

Comment (dbjoa) [reply foundationdb/pom.xml]: @busbey , No problem. I've added the skip properties for JDK10 and JDK11 to the updated PR.

**Comment by busbey** [comment]
Thanks for getting this done!

**Pull request merged**